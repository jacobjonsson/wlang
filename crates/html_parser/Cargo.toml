[package]
name = "html_parser"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
html_lexer = { path = "../html_lexer" }
html_ast = { path = "../html_ast" }

[dev-dependencies]
insta = { version = "1.8" }
