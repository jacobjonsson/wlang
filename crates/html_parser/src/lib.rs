use html_ast::{Element, Node, NodeRef};
use html_lexer::{HtmlLexer, Token};

pub struct HtmlParser {
    lexer: HtmlLexer,
    fragment: NodeRef,
    stack: Vec<NodeRef>,
    token: Token,
}

impl HtmlParser {
    pub fn new(source: &str) -> HtmlParser {
        let fragment = NodeRef::new(Node::Element(Element {
            name: "fragment".into(),
            attributes: Vec::new(),
            children: Vec::new(),
        }));

        HtmlParser {
            lexer: HtmlLexer::new(source),
            stack: vec![fragment.clone()],
            fragment,
            token: Token::EOF,
        }
    }

    pub fn parse(mut self) -> NodeRef {
        self.next();
        while self.token != Token::EOF {
            self.handle_token();
        }

        return self.fragment;
    }

    fn current(&mut self) -> NodeRef {
        self.stack.last().unwrap().clone()
    }

    fn next(&mut self) -> &Token {
        self.token = self.lexer.next();
        &self.token
    }

    fn handle_token(&mut self) {
        match self.token {
            Token::Character(ch) if ch.is_whitespace() => {
                self.next();
            }
            Token::Character(_) => self.handle_text(),
            Token::Doctype => panic!("Does not support doctype"),
            Token::Tag { .. } => self.handle_tag(),
            Token::EOF => return,
        };
    }

    fn handle_tag(&mut self) {
        let parent = self.current();
        let name = self.token.tag_name();
        let mut element = Node::new_element(name.clone());
        for attribute in self.token.attributes() {
            element.set_element_attribute(attribute.name.clone(), attribute.value.clone());
        }

        if self.token.is_end_tag() {
            self.stack.pop();
            self.next();
            return;
        }

        let element_ref = NodeRef::new(element);

        if self.token.self_closing() {
            // We don't push it onto the stack if it's self closing.
            parent.borrow_mut().append_child(element_ref);
        } else {
            parent.borrow_mut().append_child(element_ref.clone());
            self.stack.push(element_ref.clone());
        }
        self.next();
    }

    fn handle_text(&mut self) {
        let mut text = Node::new_text();
        text.append_text(self.token.character());

        while let Token::Character(ch) = self.next() {
            text.append_text(*ch);
        }
        println!("{:?}", self.token);
        self.current().borrow_mut().append_child(NodeRef::new(text));
    }
}
