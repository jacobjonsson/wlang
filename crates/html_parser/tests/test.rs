use html_parser::HtmlParser;
use insta::assert_json_snapshot;

#[test]
fn basic_test_cases() {
    assert_json_snapshot!(HtmlParser::new("abc").parse());
    assert_json_snapshot!(HtmlParser::new("<div />").parse());
    assert_json_snapshot!(HtmlParser::new("<div></div>").parse());
    assert_json_snapshot!(HtmlParser::new("<div><div><h1></h1></div></div>").parse());
    assert_json_snapshot!(HtmlParser::new("<div><div></div></div><div><div></div></div>").parse());
    assert_json_snapshot!(HtmlParser::new(
        "
        <div>
            <h1>Hello world</h1>
        </div>
    "
    )
    .parse());

    assert_json_snapshot!(
        HtmlParser::new("<div><link href=\"https://www.google.com\" /></div>").parse()
    );
}
