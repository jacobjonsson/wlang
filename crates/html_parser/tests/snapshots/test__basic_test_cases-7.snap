---
source: crates/html_parser/tests/test.rs
expression: "HtmlParser::new(\"<div><link href=\\\"https://www.google.com\\\" /></div>\").parse()"

---
{
  "Element": {
    "name": "fragment",
    "attributes": [],
    "children": [
      {
        "Element": {
          "name": "div",
          "attributes": [],
          "children": [
            {
              "Element": {
                "name": "link",
                "attributes": [
                  {
                    "name": "href",
                    "value": "https://www.google.com"
                  }
                ],
                "children": []
              }
            }
          ]
        }
      }
    ]
  }
}
